bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ClientLibrariesImpl.java

```diff
@@ -272,59 +282,91 @@ private String getInline(LibraryType libraryType) {
+                        categories.add(category);
+                    }
+                }
             }
+            return categories;
+        } catch (LoginException e) {
+            LOG.error("Cannot login as a service user", e);
+            return Collections.emptySet();
         }
-        for (String resourceType : allResourceTypes) {
-            Resource componentResource = Utils.getResource(resourceType, request, resolverFactory);
-            addClientLibraries(componentResource, libraries);
+    }
+
+    /**
+     * Gets all of the client libraries.
+     *
+     * @param resourceResolver The resource resolver.
+     * @return Set of all client libraries.
+     */
+    @NotNull
+    private Set<ClientLibrary> getAllClientLibraries(@NotNull final ResourceResolver resourceResolver) {
+        Map<String, ClientLibrary> allLibraries = htmlLibraryManager.getLibraries();
+        Set<ClientLibrary> clientLibraries = new LinkedHashSet<>();
+        for (String resourceType : getAllResourceTypes(resourceResolver)) {
+            Resource resource = resourceResolver.getResource(resourceType);
+            if (resource != null) {
+                clientLibraries.addAll(getClientLibraries(resource, allLibraries));
+            }
         }
+        return clientLibraries;
+    }
 
-        for (ClientLibrary library : libraries) {
-            for (String category : library.getCategories()) {
-                if (pattern != null) {
-                    if (pattern.matcher(category).matches()) {
-                        categories.add(category);
-                    }
-                } else {
-                    categories.add(category);
-                }
+    /**
+     * Get all resource types.
+     *
+     * @param resourceResolver The resource resolver.
+     * @return Set of all resource types under which to search for client libraries.
+     */
+    @NotNull
+    private Set<String> getAllResourceTypes(@NotNull final ResourceResolver resourceResolver) {
+        Set<String> allResourceTypes = new LinkedHashSet<>(resourceTypeSet);
+        if (inherited) {
+            for (String resourceType : resourceTypeSet) {
+                allResourceTypes.addAll(Utils.getSuperTypes(resourceType, resourceResolver));
             }
         }
-        return categories;
+        return allResourceTypes;
     }
 
     /**
-     * Adds client libraries to the provided collection, starting from the given resource
+     * Gets a list of client libraries, starting from the given resource
      * and diving into its descendants.
      *
      * @param resource - the given resource, which will be checked to see if it's a client library
-     * @param libraries - the provided collection of libraries to add to
+     * @param allLibraries - Map of all client libraries.
+     * @return List of client libraries for the given resource.
      */
-    private void addClientLibraries(Resource resource, Collection<ClientLibrary> libraries) {
```

Comment:
Same as above, we used `add*` to avoid creating new collection/set on each recursion.

Reply:
How about if we remove the recursion entirely. 
This re-introduces streams, but could be done without them. It just feels natural to me to use a stream for filtering a collection:

```JAVA
    @NotNull
    private static List<ClientLibrary> getClientLibraries(@org.jetbrains.annotations.Nullable final Resource resource,
                                                          @NotNull final Map<String, ClientLibrary> allLibraries) {
        return Optional.ofNullable(resource)
            .map(Resource::getPath)
            .map(path -> allLibraries.entrySet().stream()
                .filter(entry -> entry.getKey().equals(path) || entry.getKey().startsWith(path + "/"))
                .map(Map.Entry::getValue)
                .collect(Collectors.toList()))
            .orElseGet(Collections::emptyList);
    }
```

Since we never add a client library unless it exists in the `allLibraries` collection; why not just filter that collection?
I think that filtering an existing collection will perform better than even a small tree-traversal followed by a collection lookup. This also safe-guards against super expensive traversals that would result if there is a huge tree under the provided resource.

Reply:
As a side note, this is an example of where I personally enjoy streams.
A different style of coding, but the way I would write this would be to return the stream directly, and just `flatMap` it in '#getAllClientLibraries()', which would avoid the collection until the final results.

Reply:
Also, another thought.
Do we want to filter out clientlibs that don't allow proxy?
Since we are searching /apps or /libs for client libraries and then rendering them inline, we are effectively proxying them.

If we want to do that then it's just adding a
```JAVA
    .filter(ClientLibrary::allowProxy)
```

Reply:
Filtering `allLibraries` might be actually better, you're right. I would not worry about `allowProxy` as embedding a non-proxied clientlib in a proxied one will also make it available, similarly to what we are doing.